FOCUS #190: Numeric Format Attribute

tobrien opened the pull request:
This Pull Request contains proposed language for the numeric format attribute.   Note that this also proposes the ability to define more granular numeric types.    

Numeric
Integer
Decimal

There is also an option to add a "Data Precision" constraint to a metric to define precision over a defined set or precision specifiers.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+# Date/Time Format
```
```suggestion
# Numeric Format
```

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+
+All columns capturing a numeric value, defined in the FOCUS specification, MUST be a real number represented as an integer, a decimal value, or a value represented in scientific notation as defined in this section.
+
+## Attribute ID
```
I would consider putting Numeric Format, Numeric Precision and Examples above this line (maybe under sub headings (H3 or lower) so we don't mix up the descriptive stuff with the spec items that start with this Attribute ID header. Feels like Numeric Format being the first section and then precision and then examples make sense to me).

I think there is a normative sentence in the Precision section that we can probably move into the above normative section (line 5)

tobrien replied:
I rearranged the sections.

udam-f2 reviewed (commented):
Awesome to see this.. Thanks for putting the time and effort on this!

ljadvey commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+
+## Requirements
+
+* Metrics with a Numeric format MUST contain a single value.
+* The single value captured in metrics with a Numeric format MUST be a real, rational.
```
Agree with this

tobrien replied:
I changed it to:

* Columns with a Numeric format MUST contain a single, numeric value.
* Numeric values MUST be expressed as integer or decimal values. Fractional notation is NOT allowed.

The "real, rational number" requirement doesn't need to be there, and I think it was just distracting.  If we don't allow fractions, and we specify integer or decimal values then we're only talking about real and rational numbers.

Also, I took the line item out about arrays.  If the first line item states "single, numeric value" then we don't have to explicitly disallow arrays or ranges.

ljadvey commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+## Examples of Numeric Values Meeting Requirements
+
+* Values Meeting Numeric Requirements:
+  * "-100.2"
+  * "0"
+  * "1"
+  * "1.234"
+  * "1.23342 x 10^7"
```
I read it as such. Double quotes shouldn’t be used.

ljadvey commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+
+## Numeric Precision
+
+The FOCUS specification does not require a specific level of precision for numeric values.  The level of precision required for a given metric is determined by the provider unless specified in the metric specification.  The level of precision required for a given metric SHOULD be documented in the metric definition.  The following table outlines the options for specifying the level of precision f ...
```
Difficult…. Prefer not to define our precision but following the above on REAL number.

ljadvey commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+
+Columns that provide numeric values conforming to specified rules and formatting requirements ensure clarity, accuracy, and ease of interpretation for both humans and systems.
+
+All columns capturing a numeric value, defined in the FOCUS specification, MUST be a real number represented as an integer, a decimal value, or a value represented in scientific notation as defined in this section.
```
Scientific notation is not compute friendly…. I would prefer we don’t allow this. Thoughts?

tobrien replied:
Yeah, I could go both ways.  I think I'm leaning towards us not stating anything about scientific notation.   I just added this "just in case," but I don't see a need for it.

@ljadvey I'm going to take it out of this request - it's a distraction.

tobrien replied:
Just to make it clear, I don't know if I'm going to expressly prohibit it, but I think we just leave it out.    From a practical perspective, if I have floating point numbers, every language I know of is going to know how to parse them, but I just don't think we need to state this one way or the other.

The only place I can think of where something like this is useful, is when we have several underutilized storage  - if you are being charged in PBs, but you are only consuming a fraction of a percent of that unit, you end up with numbers like 0.000000342.   But, I don't think it's worth diving into this level of specificity on the format

ljadvey commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,104 @@
+| Integer        | Short                | 16-bit signed two's complement integer | -32,767 to +32,767 |
+| Integer        | Long                 | 32-bit signed two's complement integer | -2,147,483,647 to +2,147,483,647 |
+| Integer        | Extended             | 64-bit signed two's complement integer *or higher* | extended |
+| Float          | (unspecified)        | 64-bit binary format IEEE 754-2008 floating-point | 16 |
+| Float          | Single               | 32-bit binary format IEEE 754-2008 floating-point | 9 |
+| Float          | Double               | 64-bit binary format IEEE 754-2008 floating-point | 16 |
+| Float          | Extended             | 128-bit binary format IEEE 754-2008 floating-point *or higher* | extended |
```
Quantity would probably have integer….? But, we need to look at it case by case.

tobrien replied:
Decimal is the same as float from a standards perspective, I'm getting more specific here and just getting rid of the word float all together.   I'll reference the decimal32,64,128 to get more specific - look for the next commit to fix this.

tobrien commented:
@ljadvey  @udam-f2  - I've updated the content based on feedback from the PR.

* Removed Quotes from Examples
* Simplified the requirements
* Got rid of any mention of scientific notation (also removed the example)
* Added language that stated that providers should define data precision, but removed any language related to precision being set in the standard.
* Identifying a ISO/IEC standard to root the integer precisions values in ISE/IEC 9899:2018
* Changed all mention of the term "float" to "decimal" and referencing the decimal32,64,128 formats in IEEE 754

In general, this revision should be more concise.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+
+* Columns with a Numeric format MUST contain a single, numeric value.
+* Numeric values MUST be expressed as integer or decimal values. Fractional notation is NOT allowed.
+* Numerical values expressed with mathematical symbols, operators, or exponent values are not permitted.
```
I assume this is intentionally written as Numerical and not Numeric?

tobrien replied:
Yep, that was a mistake.  Fixed in the latest push to this branch.  thanks for catching

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+## Requirements
+
+* Columns with a Numeric format MUST contain a single, numeric value.
+* Numeric values MUST be expressed as integer or decimal values. Fractional notation is NOT allowed.
```
```suggestion
* Numeric values MUST be expressed as integer or decimal values. Fractional notation MUST NOT be used.
```

udam-f2 replied:
NOT is not a part of https://www.rfc-editor.org/rfc/rfc2119.txt. We can use SHALL NOT if that's preferable here.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+
+* Columns with a Numeric format MUST contain a single, numeric value.
+* Numeric values MUST be expressed as integer or decimal values. Fractional notation is NOT allowed.
+* Numerical values expressed with mathematical symbols, operators, or exponent values are not permitted.
```
```suggestion
* Numerical values MUST NOT be expressed with mathematical symbols, operators, or exponent values.
```

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+* Numerical values expressed with mathematical symbols, operators, or exponent values are not permitted.
+* Numeric values MUST NOT contain qualifiers or additional characters (e.g. currency symbols, units of measure, etc.).
+* Numeric values MUST NOT contain commas or punctuation marks except for a single decimal point if required to express a decimal value.
+* Numeric values MUST NOT include a character to represent sign for a positive value.  The only sign permitted is a negative sign (-) to indicate a negative value.
```
Not a 100% sure if one of the BCP-14 values should be used for the second sentence. 'MAY' is optional but in the case of negative values, that's not optional So would we say "Negative sign (-) MUST be used to indicate a negative value."?

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+| Decimal          | Double               | 64-bit binary format IEEE 754-2008 floating-point (decimal64) | 16 |
+| Decimal          | Extended      | 128-bit binary format IEEE 754-2008 floating-point (decimal128) or higher | 36+ |
+
+## Numeric Format Definitions
+
+Numeric data types are defined in each metric specification.  The following table outlines the numeric data types defined in the FOCUS specification.
+
+| Numeric Type    |                 |
+| :-------------- | :-------------- |
+| Numeric         | General Numeric Type  | Specifies any numeric value compliant with this attribute definition.  This type is used when the metric definition does not specify a more specific numeric type. |
+| Integer         | Integer               | Specifies a numeric value represented by a whole number or by zero. |
+| Decimal         | Decimal               | Specifies a numeric value represented by a decimal number |
+
+The numeric Data type is specified in the metric definition as the "Data type" constraint as shown in the following example table.
+
+|    Constraint   |      Value      |
+|:----------------|:----------------|
+| Column required | True            |
+| Data type       | Decimal         |
+| Allows nulls    | False           |
+| Value format    | Numeric value   |
+
```
I'm a little confused by this section. Do we need this? What are the possible choices for Data type and Value format in the above table? Are both necessary or should Value format just be Numeric value and Data type just say Decimal | Integer?

Also feels like this should go above the precision section

tobrien replied:
Here's what I'm trying to say in plain-speak - "You can list Numeric, Integer, or Decimal.   If you list Numeric, you are basically saying - could be either. The spec has no opinion."

Do we want to just make it require to specify either Integer or Decimal?  I'm good with that if it makes it more straightforward.

tobrien replied:
I reworded things a bit - see if it makes sense in the latest push.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+* Values Meeting Numeric Requirements:
+  * -100.2
+  * 0
+  * 1
```
since we say above not to use it for binary values, should we just suggest a different value for either line 69 or 70? e.g. 103

tobrien replied:
Done. Updated

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+* Numeric values MUST NOT include a character to represent sign for a positive value.  The only sign permitted is a negative sign (-) to indicate a negative value.
+* Numeric values MUST NOT be used to represent binary values (e.g. 0 or 1).
+
+## Numeric Precision
```
Thoughts on this and the two sections below being a sub-heading and not the same level as the standard headers we use for the sepc content? For example we could park these as H3s after the normative content at the very top (before ## Attribute ID).

tobrien replied:
ok, I rearranged content in the latest push to the branch.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,91 @@
+
+## Numeric Precision
+
+The FOCUS specification does not require a specific level of precision for numeric values.  The level of precision required for a given column is determined by the provider and SHOULD be documented as a part of a data definition published by the provider.  The following table outlines the options for specifying the level of precision for a given metric in a data definition document.
```
If we have to use BCP-14 normative content, it would be best if we can limit it to the requirements section so everything is in one place.

tobrien replied:
Ok, the precision section is an interesting case because it's normative that providers SHOULD publish this, but I still think it's not an enforceable requirement.   I rearranged everything that is a MUST up top and left this at the end.   WDYT?

flanakin commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,87 @@
+
+## Numeric Precision
+
+The FOCUS specification does not require a specific level of precision for numeric values.  The level of precision required for a given column is determined by the provider and SHOULD be documented as a part of a data definition published by the provider.  The following table outlines the options for specifying the level of precision for a given column in a data definition document.
```
nit
```suggestion
FOCUS does not require a specific level of precision for numeric values.  The level of precision required for a given column is determined by the provider and SHOULD be documented as a part of a data definition published by the provider.  The following table outlines the options for specifying the level of precision for a given column in a data definition document.
```

ijurica commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,72 @@
+
+Columns that provide numeric values conforming to specified rules and formatting requirements ensure clarity, accuracy, and ease of interpretation for humans and systems. The FOCUS specification does not require a specific level of precision for numeric values. The level of precision required for a given column is determined by the provider and should be part of a data definition published by the ...
+
+All columns capturing a numeric value, defined in the FOCUS specification, MUST follow the formatting requirements listed below. Provider generated date/time related columns SHOULD adopt the same format requirements over time.
```
```suggestion
All columns capturing a numeric value, defined in the FOCUS specification, MUST follow the formatting requirements listed below. Provider generated numeric value capturing columns SHOULD adopt the same format requirements over time.
```

ijurica replied:
except for this small oversight, great job!

ijurica commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,72 @@
+  * $32 - contains a currency symbol
+  * 3,432,342 - contains a comma
+  * +333 - contains a positive sign
+
```
E notation is prohibited or? I suggest including it as one of the examples.

marc-perreaut replied:
I understand from the requirements that the E/e exponential notation is disallowed.
This notation is supported by JSON, Python and many languages.
This notation is used in Azure billing data.
What is the rationale for disallowing it?

tobrien replied:
I wasn't aware that Azure allowed this.  @marc-perreaut - I just committed language to allow it, and to specify that values must adhere to the E notation format.  Thanks for catching this.

flanakin commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,73 @@
+* Numeric values MUST be expressed as integer or decimal values. Fractional notation MUST NOT be used.
+* Numeric values MUST NOT be expressed with mathematical symbols, operators, or exponent values.
+* Numeric values MUST NOT contain qualifiers or additional characters (e.g., currency symbols, units of measure, etc.).
+* Numeric values MUST NOT contain commas or punctuation marks except for a single decimal point if required to express a decimal value.
```
Should we be explicit about the "decimal point" being a period given regional variations?
```suggestion
* Numeric values MUST NOT contain commas or punctuation marks except for a single decimal point (`.`) if required to express a decimal value.
```

marc-perreaut commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,73 @@
+
+Columns that provide numeric values conforming to specified rules and formatting requirements ensure clarity, accuracy, and ease of interpretation for humans and systems. The FOCUS specification does not require a specific level of precision for numeric values. The level of precision required for a given column is determined by the provider and should be part of a data definition published by the ...
+
+All columns capturing a numeric value, defined in the FOCUS specification, MUST follow the formatting requirements listed below. Provider generated numeric value capturing columns SHOULD adopt the same format requirements over time.
```
>  Provider generated numeric value capturing columns SHOULD adopt the same format requirements over time.

I would propose a less strong statement:

>  Provider generated numeric value capturing columns SHOULD adopt *backward-compatible* format requirements over time.

tobrien replied:
I implemented this suggestion.  Thanks

joshk0 reviewed (changes requested):
looks good.
I wonder if we should prescribe wire formats when advanced formats (parquet, avro etc.) provide numbers?
For example, we've cooked ourselves as a data platform using FLOAT64 in BigQuery (IEEE 754 style), when we instead should have used NUMERIC. This is especially noticeable when numbers are very small; such as the unit cost of 1 byte of RAM for one hour in a compute service ($0.00000001 type number) If you take the float64 and multiply it over many hours, you might get a wrong result.

It may be out of scope for this spec though which I totally get. Just always want to find ways for people to avoid shooting themselves in the foot where I have before.

tobrien commented:
> looks good. I wonder if we should prescribe wire formats when advanced formats (parquet, avro etc.) provide numbers? For example, we've cooked ourselves as a data platform using FLOAT64 in BigQuery (IEEE 754 style), when we instead should have used NUMERIC. This is especially noticeable when numbers are very small; such as the unit cost of 1 byte of RAM for one hour in a compute service ($0.00000001 type number) If you take the float64 and multiply it over many hours, you might get a wrong result.
> 
> It may be out of scope for this spec though which I totally get. Just always want to find ways for people to avoid shooting themselves in the foot where I have before.

@joshk0 - Good evening Mr. Kwan.  I was thinking about this, and how to put it into the column specification, I don't know if it makes sense to talk about the wire format in the attribute, but I do think something needs to be said somewhere.   Can we leave it out of this format specifier and figure out another place to put it.

Example, I think we need to think about maybe another document that is focused on general tips for implementation that can go into arbitrary precision data types in different languages and databases.   Does that make sense?   So, I'm marking it as resolved for this PR, but I do think a real resolution would be to drop it into a discussion section.

udam-f2 commented on specification/attributes/numeric_format.md:
```diff
@@ -0,0 +1,76 @@
+
+Columns that provide numeric values conforming to specified rules and formatting requirements ensure clarity, accuracy, and ease of interpretation for humans and systems. The FOCUS specification does not require a specific level of precision for numeric values. The level of precision required for a given column is determined by the provider and should be part of a data definition published by the ...
+
+All columns capturing a numeric value, defined in the FOCUS specification, MUST follow the formatting requirements listed below. Provider-generated numeric value capturing columns SHOULD adopt backward-compatible format requirements over time.
```
In other attributes, we're going with language like this: 
"All columns capturing a date/time value, defined in the FOCUS specification, MUST follow the formatting requirements listed below. Provider-generated, date/time-related columns SHOULD adopt the same format requirements over time."

Should we make the others look like this OR make this consistent with the others? Fine either way. I don't feel like we need to explicitly say backward-compatible FWIW. How providers solve these issues for their customers is up to them. They might create new data sets for all we know.

marc-perreaut replied:
I think I proposed the backward-compatibility when reviewing the PR, because the original statement looked too strong to me. Sorry, I had not cross checked the language used in other language attributes. I do not have a strong opinion about stating explicitly backward compatibility and I am fine to make this consistent with others.

udam-f2 reviewed (commented):
LGTM other than a minor comment.

joshk0 reviewed (approved):
Let's figure out how to make a mention of precision somewhere later, but that's an improvement over everything here which is already good and should just go in.

udam-f2 reviewed (approved):
LGTM
